src/drivers/src/digest.rs

```diff
@@ -0,0 +1,140 @@
+use core::cell::{RefCell, Cell};
+use hil::digest::{DigestEngine, DigestMode};
+use main::{AppId, AppSlice, Container, Driver, Shared, SyscallError};
+
+/// Per-application driver data.
+pub struct AppData {
+    /// Buffer where data to be hashed will be read from.
+    input_buffer: Option<AppSlice<Shared, u8>>,
+    /// Buffer where the digest will be written to when hashing is finished.
+    output_buffer: Option<AppSlice<Shared, u8>>,
+}
+
+impl Default for AppData {
+    fn default() -> AppData {
+        AppData {
+            input_buffer: None,
+            output_buffer: None,
+        }
+    }
+}
+
+pub struct DigestDriver<'a, E: DigestEngine + 'a> {
+    engine: RefCell<&'a mut E>,
+    apps: Container<AppData>,
+    current_user: Cell<Option<AppId>>,
+}
+
+impl<'a, E: DigestEngine + 'a> DigestDriver<'a, E> {
+    pub fn new(engine: &'a mut E, container: Container<AppData>) -> DigestDriver<'a, E> {
+        DigestDriver {
+            engine: RefCell::new(engine),
+            apps: container,
+            current_user: Cell::new(None),
+        }
+    }
+}
+
+impl<'a, E: DigestEngine> Driver for DigestDriver<'a, E> {
+    fn command(&self, minor_num: usize, r2: usize, caller_id: AppId) -> isize {
+        match minor_num {
+                // Initialize hash engine (arg: digest mode)
+                0 => {
+                    self.apps
+                        .enter(caller_id, |_app_data, _| {
+                            if self.current_user.get().is_some() {
+                                return Err(SyscallError::ResourceBusy);
+                            }
+                            self.current_user.set(Some(caller_id));
+
+                            let digest_mode = match r2 {
+                                0 => DigestMode::Sha1,
+                                1 => DigestMode::Sha256,
+                                _ => return Err(SyscallError::InvalidArgument),
+                            };
+
+                            try!(self.engine.borrow_mut().initialize(digest_mode));
+
+                            Ok(0)
+                        })
+                        .unwrap_or(Err(SyscallError::Unknown))
```

Comment:
It's not so much the use of `Option` (which can be readily converted into a `Result` using `ok_or`), but the idiom of using `map` with `TakeCell`. The lambda nesting means you end up with non-flat control flow and types like `Result<Option<Result<T, E1>>, E2>` which end being actually pretty painful to collapse down to the `Result<T, E2>` you want. (As an example. I did run into this issue before when using `TakeCell` but don't remember the exact type.)

There's nothing that makes this unsolvable, I just didn't consider the burden of introducing a new cell type when `RefCell` is already in core and effectively has essentially the same semantics a worthwhile trade-off. But you're right, this is probably be a discussion for the mailing list.

I've started fixing this code to remove the panics, just didn't have time to finish it up today.